pkg/admission/reservedmetadata/admission.go

```diff
@@ -0,0 +1,117 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package reservedmetadata
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"reflect"
+	"strings"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apiserver/pkg/admission"
+	"k8s.io/apiserver/pkg/authentication/user"
+	"k8s.io/utils/strings/slices"
+)
+
+const (
+	PluginName = "apis.kcp.dev/ReservedMetadata"
+)
+
+var (
+	annotationAllowList = []string{
+		"scheduling.kcp.dev/placement",
+		"kcp.dev/namespace-locator",
+	}
+
+	labelAllowList = []string{
+		"experimental.workload.kcp.dev/scheduling-disabled",
+		"state.internal.workload.kcp.dev/*",
+	}
+)
+
+// Register registers the reserved metadata plugin for creation and updates.
+// Deletion and connect operations are not relevant as not object changes are expected here.
+func Register(plugins *admission.Plugins) {
+	plugins.Register(PluginName,
+		func(_ io.Reader) (admission.Interface, error) {
+			return &reservedMetadata{
+				Handler: admission.NewHandler(admission.Create, admission.Update),
+			}, nil
+		})
+}
+
+// reservedMetadata is a validating admission plugin protecting against mutating reserved kcp metadata.
+type reservedMetadata struct {
+	*admission.Handler
+}
+
+var _ = admission.ValidationInterface(&reservedMetadata{})
+
+// Validate asserts the underlying object for changes in labels and annotations.
+// If the user is member of the "system:masters" group, all mutations are allowed.
+func (o *reservedMetadata) Validate(ctx context.Context, a admission.Attributes, _ admission.ObjectInterfaces) (err error) {
+	newMeta, err := meta.Accessor(a.GetObject())
+	if err != nil {
+		// The object we are dealing with doesn't have object metadata defined
+		// hence it doesn't have annotations to be checked.
+		return nil
+	}
+
+	oldMeta, err := meta.Accessor(a.GetOldObject())
+	if err != nil {
+		// see above
+		return nil
+	}
+
+	if slices.Contains(a.GetUserInfo().GetGroups(), user.SystemPrivilegedGroup) {
+		return nil
+	}
+
+	if err := assertViolation(newMeta.GetAnnotations(), oldMeta.GetAnnotations(), annotationAllowList); err != nil {
+		return admission.NewForbidden(a, fmt.Errorf("annotation violation: %v", err))
+	}
+
+	if err := assertViolation(newMeta.GetLabels(), oldMeta.GetLabels(), labelAllowList); err != nil {
+		return admission.NewForbidden(a, fmt.Errorf("label violation: %v", err))
+	}
+
+	return nil
+}
+
+func assertViolation(new, old map[string]string, allowList []string) error {
+	if reflect.DeepEqual(new, old) {
```

Comment:
this is not enough. We have to check per labels/annotation whether it got added/changed/removed.

Reply:
this has to compute a real delta between old and new

Reply:
done using a very naive approach (iterating over new and old)